### ISSUE TITLE
AWSCloudWatchLoggingCategoryClient asks for Current User Identifier During Configuration (CRASH)

### ISSUE THREAD

**Issue by Hartistic** [aws-amplify/amplify-swift#3897]
### Describe the bug

While configuring Amplify with an AWSCloudWatchLoggingPluginConfiguration and an AWSCognitoAuthPlugin configuration the library has a race condition that crashes the application 50% of the time with an error of `Thread 20: Fatal error: Authentication category is not configured. Call Amplify.configure() before using any methods on the category.`

The problem is that these two configurations are added to Amplify and this crash occurs when calling Amplify.configure().

Order of Methods that causes crash to occur:

1. `try configure(CategoryType.logging.category, using: resolvedConfiguration)`
2. `try configurable.configure(using: configuration)`
3. `try configure(using: categoryConfiguration(from: amplifyConfiguration))`
4. `try Amplify.configure(plugins: Array(plugins.values), using: configuration)`
5. `try plugin.configure(using: pluginConfiguration)`
6. `DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
            self.loggingClient.takeUserIdentifierFromCurrentUser()
        }`
7. ```func takeUserIdentifierFromCurrentUser() {
        Task {
            do {
                let user = try await authentication.getCurrentUser()
                self.userIdentifier = user.userId
            } catch {
                self.userIdentifier = nil
            }
            self.updateSessionControllers()
        }
    }```
8. ```public func getCurrentUser() async throws -> AuthUser {
        try await plugin.getCurrentUser()
    }```
9. ```return Fatal.preconditionFailure(
                """
                \(categoryType.displayName) category is not configured. Call Amplify.configure() before using \
                any methods on the category.
                """
            )```

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Add AWSCloudWatchLoggingPluginConfiguration and AWSCognitoAuthPlugin to Amplify:

    private func addAuthenticationPlugin() throws {
        let authPlugin = AWSCognitoAuthPlugin(networkPreferences: Self.networkPreferences)
        do {
            try Amplify.add(plugin: authPlugin)
        } catch {
            throw error
        }
    }

    private func addCloudWatchLoggingPlugin() throws {
        do {
            let logLevels: [String: LogLevel] = ["AWSAuthFetchSessionTask": .error]
            let loggingConstraints = LoggingConstraints(defaultLogLevel: .debug, categoryLogLevel: logLevels)
            let loggingConfiguration = AWSCloudWatchLoggingPluginConfiguration(logGroupName: "your log group",
                                                                               region: "us-east-1",
                                                                               localStoreMaxSizeInMB: 1,
                                                                               flushIntervalInSeconds: 10,
                                                                               loggingConstraints: loggingConstraints)
            let plugin = AWSCloudWatchLoggingPlugin(loggingPluginConfiguration: loggingConfiguration)
            try Amplify.add(plugin: plugin)
        } catch {
            throw error
        }
    }
2. load a amplifyconfiguration.json and call configure:

    private func loadAmplifyConfiguration() throws {
        do {
            let amplifyConfig = try getLocalAmplifyConfigurationFile()
            try Amplify.configure(amplifyConfig)
        } catch {
            throw error
        }
    }
3. On iPhone 15 Pro Max and newer seems to crash way more often.
```

### Expected behavior

The AWSCloudWatchLoggingPluginConfiguration shouldn't be asking for Authentication details during it's configuration...the library says you can only call configure() once, so configuring Auth and then Logging isn't possible.

### Amplify Framework Version

2.41.1

### Amplify Categories

Analytics, API, Auth

### Dependency manager

Swift PM

### Swift version

5.8.0

### CLI version

I am not sure

### Xcode version

16.0

### Is this a regression?

Yes

### Platforms

iOS

### OS Version

IOS 18

### Device

iPhone 15 Pro Max

### Specific to simulators

no

**Comment by thisisabhash** [comment]
Hello,

`AWSCloudWatchLoggingPlugin` needs Auth Plugin to be configured. You'd need to setup Amplify Auth through CLI by following steps here: https://docs.amplify.aws/gen1/swift/build-a-backend/auth/set-up-auth/

Then you can setup `AWSCloudWatchLoggingPlugin` by following steps here: https://docs.amplify.aws/gen1/swift/build-a-backend/more-features/logging/set-up-logging/

```
import Amplify
import AWSCognitoAuthPlugin
import AWSCloudWatchLoggingPlugin

private func addCloudWatchLoggingPlugin() throws {
    do {
        let logLevels: [String: LogLevel] = ["AWSAuthFetchSessionTask": .error]
        let loggingConstraints = LoggingConstraints(defaultLogLevel: .debug, categoryLogLevel: logLevels)
        let loggingConfiguration = AWSCloudWatchLoggingPluginConfiguration(logGroupName: "your log group",
                                                                           region: "us-east-1",
                                                                           localStoreMaxSizeInMB: 1,
                                                                           flushIntervalInSeconds: 10,
                                                                           loggingConstraints: loggingConstraints)
        let plugin = AWSCloudWatchLoggingPlugin(loggingPluginConfiguration: loggingConfiguration)
        try Amplify.add(plugin: plugin)
    } catch {
        throw error
    }
}

init() {
    do {
        try Amplify.add(plugin: AWSCognitoAuthPlugin())
        try addCloudWatchLoggingPlugin()
        try Amplify.configure()
    } catch {
        assert(false, "Error initializing Amplify: \(error)")
    }
}
```

**Comment by Hartistic** [comment]
Thank you so much for the quick response @thisisabhash, we are not using the CLI and Auth is already setup.   I think the main culprit seems to be this:
```
DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
            self.loggingClient.takeUserIdentifierFromCurrentUser()
        }
        ```

**Comment by Hartistic** [comment]
Absolutely @harsh62, and thank you guys so much, Im actually hoping I am just doing something out of order and that this is an easy resolution. 

We have a SwiftUI Application.

1).  On App Launch:

- We start listening to the Auth hub first (this way we know when Amplify Configuration finishes):
```
 self.unsubscribeToken = Amplify.Hub.listen(to: .auth) { payload in
      UserStateMachine.shared.handleEvent(payload)
}
 ```
 
 - Next we add all the plugins we need to Amplify:
 ```
        // Authentication:
        try addAuthenticationPlugin()
        // Notifications:
        try addNotificationPlugin()
        // Cloud Logging:
        try addCloudWatchLoggingPlugin()
        // Pinpoint Analytics:
        try addAnalyticsPlugin()
        // AWS API:
        try addAPIPlugin()
 ```
- Next we call Amplify.configure (we load our amplifyconfiguration.json file locally that has our specific settings)
```
   private func loadAmplifyConfiguration() throws {
        do {
            let amplifyConfig = try self.loadConfiguration() // Loads JSON file
            try Amplify.configure(amplifyConfig)
        } catch {
            throw error
        }
    }
```
- Once try await for configuring is done we immediately call a force refresh:
```
    /// Check The Auth Session:
    func forceRefreshAuthSession() async throws {
        let plugin = try getAWSCognitoAuthPlugin()
        let session = try await plugin.fetchAuthSession(options: .forceRefresh())
        print("::: AUTH PLUGIN: AMPLIFY USER SESSION IS SIGNED IN - \(session.isSignedIn)")
    }
```

**Comment by Hartistic** [comment]
@harsh62 absolutely, I cannot share it on this thread, could I share it with you privately somehow?

**Comment by Hartistic** [comment]
> Got it and why do you suspect that the following snippet is a problem?
> 
> ```
> DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
>      self.loggingClient.takeUserIdentifierFromCurrentUser()
> }
> ```
> 
> From what I see is that the Auth plugin is being configured before the logging plugin, so this snippet should work, because Auth is already configured.

@harsh62 my crash backtraces lead back to that code block.

**Comment by Hartistic** [comment]
@harsh62 my Slack is: `joshua.hart@dfinitiv.io` if that helps

**Comment by Hartistic** [comment]
@harsh62 it definitely seems like a race condition.  Again way more prevalent on the newer devices (try on the iPhone 15 Pro Max or 16 Pro Max).  If we remove the AWSCloudWatchLoggingPluginConfiguration plugin, no crashes occur and everything runs smoothly.  I will try and make a demo project that contains the crash when I get some available time.  I do think from our perspective, this code should have some fail safes in it:

```
DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
     self.loggingClient.takeUserIdentifierFromCurrentUser()
}
```

It does seem to get first before Auth is configured.

**Comment by Hartistic** [comment]
> @harsh62 it definitely seems like a race condition. Again way more prevalent on the newer devices (try on the iPhone 15 Pro Max or 16 Pro Max). If we remove the AWSCloudWatchLoggingPluginConfiguration plugin, no crashes occur and everything runs smoothly. I will try and make a demo project that contains the crash when I get some available time. I do think from our perspective, this code should have some fail safes in it:
> 
> ```
> DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
>      self.loggingClient.takeUserIdentifierFromCurrentUser()
> }
> ```
> 
> It does seem to get first before Auth is configured.

Perhaps this is ran after the Auth Configuration is **confirmed** configured in the library?
